SwrveSDK/src/amazon/java/com/swrve/sdk/adm/SwrveAdmIntentService.java

```diff
@@ -0,0 +1,254 @@
+
+import com.amazon.device.messaging.ADMMessageHandlerBase;
+import com.google.gson.reflect.TypeToken;
+import com.swrve.sdk.ISwrveBase;
+import com.swrve.sdk.Swrve;
+import com.swrve.sdk.SwrveHelper;
+import com.swrve.sdk.SwrveLogger;
+import com.swrve.sdk.SwrvePushEngageReceiver;
+import com.swrve.sdk.SwrveSDK;
+import com.swrve.sdk.qa.SwrveQAUser;
+
+import com.google.gson.Gson;
+
+import java.util.Date;
+import java.util.Iterator;
+import java.util.LinkedList;
+
+public class SwrveAdmIntentService extends ADMMessageHandlerBase {
+    private final static String TAG = "SwrveAdm";
+    private final static String AMAZON_RECENT_PUSH_IDS = "recent_push_ids";
+    private final static String AMAZON_PREFERENCES = "swrve_amazon_pref";
+    private final int MAX_ID_CACHE_ITEMS = 16;
+
+    public SwrveAdmIntentService() {
+        super(SwrveAdmIntentService.class.getName());
+    }
+
+    public SwrveAdmIntentService(final String className) {
+        super(className);
+    }
+
+    @Override
+    protected void onMessage(final Intent intent) {
+        if (intent == null) {
+            SwrveLogger.e(TAG, "Unexpected null intent");
+            return;
+        }
+
+        final Bundle extras = intent.getExtras();
+        if (extras != null && !extras.isEmpty()) {  // has effect of unparcelling Bundle
+            SwrveLogger.i(TAG, "Received ADM notification: " + extras.toString());
+            processRemoteNotification(extras);
+        }
+    }
+
+    @Override
+    protected void onRegistrationError(final String string) {
+        //This is considered fatal for ADM
+        SwrveLogger.e(TAG, "ADM Registration Error. Error string: " + string);
+    }
+
+    @Override
+    protected void onRegistered(final String registrationId) {
+        SwrveLogger.i(TAG, "ADM Registered. RegistrationId: " + registrationId);
+        ISwrveBase sdk = SwrveSDK.getInstance();
+        if (sdk != null && sdk instanceof Swrve) {
+            ((Swrve) sdk).onRegistrationIdReceived(registrationId);
+        } else {
+            SwrveLogger.e(TAG, "Could not notify the SDK of a new token. Consider using the shared instance.");
+        }
+    }
+
+    @Override
+    protected void onUnregistered(final String registrationId) {
+        SwrveLogger.i(TAG, "ADM Unregistered. RegistrationId: " + registrationId);
+    }
+
+    protected boolean isSwrveRemoteNotification(final Bundle msg) {
+        Object rawId = msg.get(SwrveAdmConstants.SWRVE_TRACKING_KEY);
+        String msgId = (rawId != null) ? rawId.toString() : null;
+        return !SwrveHelper.isNullOrEmpty(msgId);
+    }
+
+    private void processRemoteNotification(Bundle msg) {
+        if (!isSwrveRemoteNotification(msg)) {
+            SwrveLogger.i(TAG, "ADM notification: but not processing as it doesn't contain: " + SwrveAdmConstants.SWRVE_TRACKING_KEY);
+            return;
+        }
+
+        //Get the context
+        Context context = getApplicationContext();
```

Comment:
Taking lead from similar scenario in service: https://github.com/Swrve/swrve-android-sdk/blob/release-4_7/SwrveSDK/src/google/java/com/swrve/sdk/gcm/SwrveGcmIntentService.java#L21. Change? Worried there were subtleties involved. Could change to "this.